### PR TITLE
Rename functions managing special weapons to improve readability for engine devellopper

### PR DESCRIPTION
**Pull request by newfrenchy83**
When I was looking at the special weapons extension to [abilities] tags, I didn't understand the concept of lists at first, until someone added the .append function to add a list to a list. other, but I kept the names I used for the functions. Now that this work is completed, I would like the function calls to find their original names (get_specials and get_special_bool) among other things to clarify the reading for a developer.

### CODE REVIEW

**Comment by Pentarctagon** [comment]
Adding "_old" to a method name generally isn't a good way to improve readability unless there's some doxygen documentation to go with it that explains what that means and why it's the "old" way of doing something.

**Comment by newfrenchy83** [comment]
> 
> 
> Adding "_old" to a method name generally isn't a good way to improve readability unless there's some doxygen documentation to go with it that explains what that means and why it's the "old" way of doing something.

i will replace "old" by "temp"

**Comment by newfrenchy83** [comment]
@Pentarctagon i will use"without_abilities" instead

**Comment by CelticMinstrel** [comment]
So "_without_abilities" means it doesn't check if someone is affected by a leadership-like ability that would grant that tag? Why does that even exist?

**Comment by newfrenchy83** [comment]
> 
> 
> So "_without_abilities" means it doesn't check if someone is affected by a leadership-like ability that would grant that tag? Why does that even exist?

the function _without_abilities are the old function who check specials in [specials] renamed for what get_specials and get_special_bool point the globals specials function with check both abilities and specials tags and if the code don't writed in one function, this because case of function without abilities are called in the code (attack_type.cpp and special_id/type  without active, or the plague special in attack.cpp, or in ai where simple_check=true and not compatible with abilities)

**Comment by CelticMinstrel** [comment]
> this because case of function without abilities are called in the code (attack_type.cpp and special_id/type without active, or the plague special in attack.cpp, or in ai where simple_check=true and not compatible wit

Why?

**Comment by newfrenchy83** [comment]
i replace "without_abilities" by "_temp" and i midify code for these function don't called outside of get_specials and get_special_bool

**Comment by Pentarctagon** [comment]
I mean, "_temp" isn't really any better than "_old".

**Comment by newfrenchy83** [comment]
> 
> 
> I mean, "_temp" isn't really any better than "_old".

have you suggestion?

**Comment by Pentarctagon** [comment]
I'd say that if the method checks just specials, then it should have "specials" in the name.  If it checks specials and abilities, then it should have "specials_and_abilities" in the name.

**Comment by newfrenchy83** [comment]
> 
> 
> I'd say that if the method checks just specials, then it should have "specials" in the name. If it checks specials and abilities, then it should have "specials_and_abilities" in the name.

it's done.

**Comment by CelticMinstrel** [comment]
I'd still like to know why there are places that check specials but not abilities…

**Comment by Pentarctagon** [comment]
As another change, I'd prefer methods that are named like `get_..._bool` be changed to `has_...`.  So for example change `get_special_and_abilities_bool` to `has_special_or_abilities`.

**Comment by newfrenchy83** [comment]
> 
> 
> As another change, I'd prefer methods that are named like `get_..._bool` be changed to `has_...`. So for example change `get_special_and_abilities_bool` to `has_special_or_abilities`.

only get_..._bool, not get... ?

**Comment by Pentarctagon** [comment]
The methods that don't end in "_bool" return special/ability information rather than a boolean, correct?

**Comment by newfrenchy83** [comment]
> 
> 
> The methods that don't end in "_bool" return special/ability information rather than a boolean, correct?

yes

**Comment by Pentarctagon** [comment]
Right, so for methods that return a boolean, it's generally better form for the method name to be phrased somewhat like a yes or no question.  As an example, in `utils/math` there's the `is_even` and `is_odd` methods, rather than something like `get_even_bool` and `get_odd_bool`.

**Comment by newfrenchy83** [comment]
@Pentarctagon done.

File: src/units/abilities.cpp [hunk 1]
```diff
@@ -1359,9 +1359,9 @@ bool attack_type::get_special_ability_bool(const std::string& special, bool spec
 	return false;
 }
 
-bool attack_type::bool_ability(const std::string& special, bool special_id, bool special_tags) const
+bool attack_type::has_special_and_abilities(const std::string& special, bool special_id, bool special_tags) const
```
Comment (soliton-) [inline src/units/abilities.cpp:1362]: Surely this should be called `has_special_or_ability`. The implementation clearly uses or.

Comment (newfrenchy83) [reply src/units/abilities.cpp]: It's done

Comment (soliton-) [reply src/units/abilities.cpp]: If you want to go with plural it should be `has_specials_or_abilities`. To me it doesn't look like this is about multiple specials or abilities though. The input is a single string.

**Pull request merged**